File: backend/services/registry.py
from __future__ import annotations

from typing import Any

from backend.services.base import BaseParcelService
from backend.services.wright import WrightParcelService
from backend.services.hennepin import HennepinParcelService
from backend.services.stlouis import StLouisParcelService
from backend.services.sherburne import SherburneParcelService
from backend.services.anoka import AnokaParcelService
from backend.services.ramsey import RamseyParcelService
from backend.services.olmsted import OlmstedParcelService
from backend.services.chisago import ChisagoParcelService
from backend.services.morrison import MorrisonParcelService
from backend.services.scott import ScottParcelService
from backend.services.aitkin import AitkinParcelService
from backend.services.koochiching import KoochichingParcelService
from backend.services.beltrami import BeltramiParcelService
from backend.services.mn_state import (
    DakotaParcelService,
    WashingtonParcelService,
    CarverParcelService,
)


COUNTY_CLASSES: dict[str, type[BaseParcelService]] = {
    "wright": WrightParcelService,
    "hennepin": HennepinParcelService,
    "stlouis": StLouisParcelService,
    "sherburne": SherburneParcelService,
    "anoka": AnokaParcelService,
    "ramsey": RamseyParcelService,
    "olmsted": OlmstedParcelService,
    "chisago": ChisagoParcelService,
    "morrison": MorrisonParcelService,
    "scott": ScottParcelService,
    "aitkin": AitkinParcelService,
    "koochiching": KoochichingParcelService,
    "beltrami": BeltramiParcelService,
    "dakota": DakotaParcelService,
    "washington": WashingtonParcelService,
    "carver": CarverParcelService,
}

COUNTY_LABELS: dict[str, str] = {
    "wright": "Wright County",
    "hennepin": "Hennepin County",
    "stlouis": "St. Louis County",
    "sherburne": "Sherburne County",
    "anoka": "Anoka County",
    "ramsey": "Ramsey County",
    "olmsted": "Olmsted County",
    "chisago": "Chisago County",
    "morrison": "Morrison County",
    "scott": "Scott County",
    "aitkin": "Aitkin County",
    "koochiching": "Koochiching County",
    "beltrami": "Beltrami County",
    "dakota": "Dakota County",
    "washington": "Washington County",
    "carver": "Carver County",
}


def create_service(county: str, **kwargs: Any) -> BaseParcelService:
    cls = COUNTY_CLASSES.get(county)
    if cls is None:
        raise ValueError(f"Unknown county: {county}. Options: {list(COUNTY_CLASSES.keys())}")
    return cls(**kwargs)

File: web/js/providers/registry.js
// Port of backend/services/registry.py
import { WrightProvider } from "./wright.js";
import { HennepinProvider } from "./hennepin.js";
import { StLouisProvider } from "./stlouis.js";
import { SherburneProvider } from "./sherburne.js";
import { AnokaProvider } from "./anoka.js";
import { RamseyProvider } from "./ramsey.js";
import { OlmstedProvider } from "./olmsted.js";
import { ChisagoProvider } from "./chisago.js";
import { MorrisonProvider } from "./morrison.js";
import { ScottProvider } from "./scott.js";
import { AitkinProvider } from "./aitkin.js";
import { KoochichingProvider } from "./koochiching.js";
import { BeltramiProvider } from "./beltrami.js";
import {
  DakotaProvider,
  WashingtonProvider,
  CarverProvider,
} from "./mn_state.js";

export const COUNTY_CLASSES = {
  wright: WrightProvider,
  hennepin: HennepinProvider,
  stlouis: StLouisProvider,
  sherburne: SherburneProvider,
  anoka: AnokaProvider,
  ramsey: RamseyProvider,
  olmsted: OlmstedProvider,
  chisago: ChisagoProvider,
  morrison: MorrisonProvider,
  scott: ScottProvider,
  aitkin: AitkinProvider,
  koochiching: KoochichingProvider,
  beltrami: BeltramiProvider,
  dakota: DakotaProvider,
  washington: WashingtonProvider,
  carver: CarverProvider,
};

export const COUNTY_LABELS = {
  wright: "Wright County",
  hennepin: "Hennepin County",
  stlouis: "St. Louis County",
  sherburne: "Sherburne County",
  anoka: "Anoka County",
  ramsey: "Ramsey County",
  olmsted: "Olmsted County",
  chisago: "Chisago County",
  morrison: "Morrison County",
  scott: "Scott County",
  aitkin: "Aitkin County",
  koochiching: "Koochiching County",
  beltrami: "Beltrami County",
  dakota: "Dakota County",
  washington: "Washington County",
  carver: "Carver County",
};

export function createService(county, options = {}) {
  const Cls = COUNTY_CLASSES[county];
  if (!Cls) {
    throw new Error(
      `Unknown county: ${county}. Options: ${Object.keys(COUNTY_CLASSES).join(", ")}`
    );
  }
  return new Cls(options);
}
